### PR TITLE
Interop between CAGRA and HNSW

### PR DESCRIPTION
**Pull request by divyegala**
Depends on https://github.com/facebookresearch/faiss/pull/3084

### CODE REVIEW

File: faiss/gpu/GpuIndexCagra.h [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    /// Lower limit of hashmap bit length. More than 8.
+    size_t hashmap_min_bitlen = 0;
+    /// Upper limit of hashmap fill rate. More than 0.1, less than 0.9.
+    float hashmap_max_fill_rate = 0.5;
+
+    /// Number of iterations of initial random seed node selection. 1 or more.
+
+    uint32_t num_random_samplings = 1;
+    /// Bit mask used for initial random seed node selection.
+    uint64_t rand_xor_mask = 0x128394;
```
Comment (mdouze) [inline faiss/gpu/GpuIndexCagra.h:235]: maybe call this seed for clarity

**Review by mdouze — COMMENTED** [review]
If possible, please add a python test as well. See 

https://github.com/facebookresearch/faiss/blob/main/faiss/gpu/test/test_gpu_index.py

for an example. Otherwise I'll do it in a follow-up diff.

**Comment by divyegala** [comment]
@mdouze I addressed your first batch of reviews. Could you re-review when it is convenient for you?

File: CMakeLists.txt [hunk 2]
```diff
@@ -46,6 +46,8 @@ project(faiss
   LANGUAGES ${FAISS_LANGUAGES})
 include(GNUInstallDirs)
 
+set(CMAKE_INSTALL_PREFIX "$ENV{CONDA_PREFIX}")
```
Comment (mdouze) [inline CMakeLists.txt:49]: Not sure we want to keep this as Faiss is not necessarily compiled in a conda env.

**Review by mdouze — COMMENTED** [review]
What is the level of parity between GPU search and CPU search? The search function on CPU does not have access to the Cagra search parameters. 

Could you write a few python tests taking inspiration from 

https://github.com/facebookresearch/faiss/blob/main/faiss/gpu/test/test_contrib_gpu.py

I would like to see: 

- build and search on GPU 

- index_gpu_to_cpu on a Cagra graph 

- serialization / deserialization 

- search on CPU

Thanks!

File: CMakeLists.txt (hunk 2 above)
Comment (divyegala) [reply CMakeLists.txt:49]: Ah, sorry. This was just for development purposes - I will remove it.

**Comment by divyegala** [comment]
@mdouze thanks for your review! Btw, the C++ tests cover all the tests you mentioned except serialization. Nonetheless, I will definitely add python tests per your suggestion.

**Comment by divyegala** [comment]
@mdouze I seem to be unable to build faiss C++ tests after merging some recent commits into my branch. This is the error I get, can you help? I have tried deleting build/cache and building again:
```
/home/nfs/dgala/faiss/build/tests/faiss_test: symbol lookup error: /home/nfs/dgala/faiss/build/tests/faiss_test: undefined symbol: _ZN5faiss15TimeoutCallback5resetEd
```

**Comment by Di-Is** [comment]
Hi @divyegala!

Can you share the steps to reproduce the error?
I have built [working branch](https://github.com/divyegala/faiss/tree/raft-cagra-hnsw) in my environment and no error occurred.

**Comment by divyegala** [comment]
@Di-Is I figured out what the issue is. I build FAISS in a conda environment, and re-configuring FAISS CMake does not overwrite the FAISS header files that get copied the very first time when I configure FAISS CMake. As a consequence, any time there's changes to an API or new APIs are added, the header files in the conda environment become stale.

**Comment by divyegala** [comment]
@mdouze I have added Python tests, could you review when possible?

**Comment by mdouze** [comment]
Thanks! Importing to the internal repo.

**Pull request closed without merging**